File: csharp/AntFfi/SelfEncryption.cs
using AntFfi.Native;

namespace AntFfi;

/// <summary>
/// Represents encrypted data returned from the self-encryption algorithm.
/// </summary>
/// <remarks>
/// This is an opaque blob containing the serialized EncryptedData record from Rust,
/// which includes the datamap chunk and content chunks needed for decryption.
/// </remarks>
public class EncryptedData
{
    /// <summary>
    /// The raw serialized encrypted data (UniFFI format).
    /// </summary>
    internal byte[] RawData { get; }

    internal EncryptedData(byte[] rawData)
    {
        RawData = rawData;
    }

    /// <summary>
    /// Gets the size of the encrypted data in bytes.
    /// </summary>
    public int Size => RawData.Length;
}

/// <summary>
/// Provides self-encryption and decryption functionality.
/// </summary>
/// <remarks>
/// Self-encryption is a content-based encryption scheme where the data is encrypted
/// using keys derived from its own content. This provides convergent encryption
/// where the same data always produces the same encrypted output.
/// </remarks>
public static class SelfEncryption
{
    /// <summary>
    /// Encrypts data using the self-encryption algorithm.
    /// </summary>
    /// <param name="data">The data to encrypt.</param>
    /// <returns>The encrypted data, which can be passed to <see cref="Decrypt"/> to recover the original data.</returns>
    /// <exception cref="EncryptionException">Thrown if encryption fails.</exception>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    public static EncryptedData Encrypt(byte[] data)
    {
        ArgumentNullException.ThrowIfNull(data);

        // Serialize data with UniFFI format (4-byte big-endian length prefix)
        var inputBuffer = UniFFIHelpers.ToRustBuffer(data);

        try
        {
            var status = RustCallStatus.Create();
            var resultBuffer = NativeMethods.Encrypt(inputBuffer, ref status);

            if (status.IsError)
            {
                throw new EncryptionException("Encryption failed", status);
            }

            // Get the raw bytes (don't deserialize - this is the serialized EncryptedData record)
            var encryptedBytes = UniFFIHelpers.FromRustBufferRaw(resultBuffer);
            return new EncryptedData(encryptedBytes);
        }
        catch (AntFfiException)
        {
            throw;
        }
        catch (Exception ex)
        {
            throw new EncryptionException($"Encryption failed: {ex.Message}");
        }
    }

    /// <summary>
    /// Encrypts a string using the self-encryption algorithm.
    /// </summary>
    /// <param name="text">The text to encrypt (will be encoded as UTF-8).</param>
    /// <returns>The encrypted data.</returns>
    public static EncryptedData Encrypt(string text)
    {
        ArgumentNullException.ThrowIfNull(text);
        return Encrypt(System.Text.Encoding.UTF8.GetBytes(text));
    }

    /// <summary>
    /// Decrypts data that was previously encrypted with <see cref="Encrypt"/>.
    /// </summary>
    /// <param name="encryptedData">The encrypted data to decrypt.</param>
    /// <returns>The original decrypted data.</returns>
    /// <exception cref="DecryptionException">Thrown if decryption fails.</exception>
    /// <exception cref="ArgumentNullException">Thrown if encryptedData is null.</exception>
    public static byte[] Decrypt(EncryptedData encryptedData)
    {
        ArgumentNullException.ThrowIfNull(encryptedData);

        // Create a RustBuffer from the raw encrypted data
        var pinnedHandle = System.Runtime.InteropServices.GCHandle.Alloc(
            encryptedData.RawData,
            System.Runtime.InteropServices.GCHandleType.Pinned);

        try
        {
            var foreignBytes = new ForeignBytes
            {
                Len = encryptedData.RawData.Length,
                Data = pinnedHandle.AddrOfPinnedObject()
            };

            var status = RustCallStatus.Create();
            var inputBuffer = NativeMethods.RustBufferFromBytes(foreignBytes, ref status);

            if (status.IsError)
            {
                throw new DecryptionException("Failed to create input buffer", status);
            }

            status = RustCallStatus.Create();
            var resultBuffer = NativeMethods.Decrypt(inputBuffer, ref status);

            if (status.IsError)
            {
                throw new DecryptionException("Decryption failed", status);
            }

            // Deserialize the result (skip UniFFI length prefix)
            return UniFFIHelpers.FromRustBuffer(resultBuffer);
        }
        catch (AntFfiException)
        {
            throw;
        }
        catch (Exception ex)
        {
            throw new DecryptionException($"Decryption failed: {ex.Message}");
        }
        finally
        {
            pinnedHandle.Free();
        }
    }

    /// <summary>
    /// Decrypts data and returns it as a UTF-8 string.
    /// </summary>
    /// <param name="encryptedData">The encrypted data to decrypt.</param>
    /// <returns>The decrypted data as a string.</returns>
    public static string DecryptToString(EncryptedData encryptedData)
    {
        var bytes = Decrypt(encryptedData);
        return System.Text.Encoding.UTF8.GetString(bytes);
    }
}

File: lua/ant_ffi/self_encryption.lua
--[[
  Self-encryption functions for ant_ffi

  This module provides encrypt and decrypt functions using
  the self-encryption algorithm from the Autonomi network.
]]

local ffi = require("ffi")

local M = {}

-- Will be set by init.lua
M._lib = nil
M._helpers = nil
M._errors = nil

-- Initialize module with dependencies
function M._init(lib, helpers, errors)
    M._lib = lib
    M._helpers = helpers
    M._errors = errors
end

--[[
  Encrypt data using self-encryption.

  Self-encryption provides content-addressable encryption where
  identical data always produces identical encrypted output.

  @param data (string) - The data to encrypt
  @return string - The encrypted data
  @raises error if encryption fails
]]
function M.encrypt(data)
    assert(M._lib, "self_encryption not initialized")
    assert(type(data) == "string", "data must be a string")

    -- Convert input to RustBuffer with UniFFI serialization
    local input_buf = M._helpers.string_to_rustbuffer(data)

    -- Call encrypt (returns RustBuffer directly)
    local status = M._errors.new_status()
    local result = M._lib.uniffi_ant_ffi_fn_func_encrypt(input_buf, status)
    M._errors.check_status(status, "encrypt")

    -- Return raw buffer contents (includes UniFFI serialization for decrypt)
    return M._helpers.rustbuffer_to_raw_string(result)
end

--[[
  Decrypt self-encrypted data.

  @param encrypted_data (string) - The encrypted data from encrypt()
  @return string - The original decrypted data
  @raises error if decryption fails
]]
function M.decrypt(encrypted_data)
    assert(M._lib, "self_encryption not initialized")
    assert(type(encrypted_data) == "string", "encrypted_data must be a string")

    -- Convert input to RustBuffer WITHOUT adding prefix (encrypted data already serialized)
    local input_buf = M._helpers.raw_string_to_rustbuffer(encrypted_data)

    -- Call decrypt (returns RustBuffer directly)
    local status = M._errors.new_status()
    local result = M._lib.uniffi_ant_ffi_fn_func_decrypt(input_buf, status)
    M._errors.check_status(status, "decrypt")

    -- Parse UniFFI serialization to get original string
    return M._helpers.rustbuffer_to_string(result)
end

--[[
  Encrypt bytes (table of byte values) using self-encryption.

  @param bytes (table) - Array of byte values (0-255) to encrypt
  @return table - Array of encrypted byte values
  @raises error if encryption fails
]]
function M.encrypt_bytes(bytes)
    assert(M._lib, "self_encryption not initialized")
    assert(type(bytes) == "table", "bytes must be a table")

    -- Convert bytes to string, encrypt, return as byte table
    local str = ""
    for i = 1, #bytes do
        str = str .. string.char(bytes[i])
    end

    local encrypted_str = M.encrypt(str)

    -- Convert encrypted string back to byte table
    local result = {}
    for i = 1, #encrypted_str do
        result[i] = string.byte(encrypted_str, i)
    end
    return result
end

--[[
  Decrypt self-encrypted bytes.

  @param encrypted_bytes (table) - Array of encrypted byte values
  @return table - Array of original decrypted byte values
  @raises error if decryption fails
]]
function M.decrypt_bytes(encrypted_bytes)
    assert(M._lib, "self_encryption not initialized")
    assert(type(encrypted_bytes) == "table", "encrypted_bytes must be a table")

    -- Convert bytes to string, decrypt, return as byte table
    local str = ""
    for i = 1, #encrypted_bytes do
        str = str .. string.char(encrypted_bytes[i])
    end

    local decrypted_str = M.decrypt(str)

    -- Convert decrypted string back to byte table
    local result = {}
    for i = 1, #decrypted_str do
        result[i] = string.byte(decrypted_str, i)
    end
    return result
end

return M
